### PR TITLE
feat(obstacle avoidance planner): fix out curve, make calculation cost low, make optimization stable, refactor, etc.

### PR DESCRIPTION
**Pull request by takayuki5168**
## Related Issue(required)

https://github.com/autowarefoundation/autoware.universe/issues/366

## Description(required)

This is a PR of the feature branch to improve obstacle avoidance planner, including
- fix out curve
    - previous: the base link (back wheel center) tries to follow the reference trajectory
    - current: the geometrical center or front wheel center (can be designated with rosparam) tries to follow the reference trajectory
- make calculation cost low
    - remove redundant implementation
    - remove publishing unnecessary topics (can be designated with rosparam)
    - Fixing out curve issue since solvability gets higher.
    - use sparse matrix
    - use Linf norm for MPT's soft constraints instead of L2 norm
    - make size of obstacle clearance map small
    - connect optimized points and remain points with spline interpolation instead of non-robust EB
    - etc..
- make optimization stable (by dealing with the fixed length well)
    - make fixing distance not so long, and remove them from optimization (previously, fixing points are also optimization's variable)
    - tune optimization
    - etc..
- refactor
    - organize main functions in node.cpp, and add flowchart in README
    - refactor utils functions
    - refactor mpt_optimizer.cpp
    - etc..
- etc

## Review Procedure(required)

try many cases with simple psim in odaiba/kashiwanoha

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Read [commit-guidelines][commit-guidelines]
- [x] Assign PR to reviewer

If you are adding new package following items are required:

- [ ] Documentation with description of the package is available
- [ ] A sample launch file and parameter file are available if the package contains executable nodes

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [x] Commits are properly organized and messages are according to the guideline
- [x] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets

## CI Checks

- **Build and test for PR / build-and-test-pr**: Required to pass before the merge.
- **Build and test for PR / clang-tidy-pr**: NOT required to pass before the merge. It is up to the reviewer(s). Found false positives? See the [guidelines][clang-tidy-guidelines].
- **Check spelling**: NOT required to pass before the merge. It is up to the reviewer(s). See [here][spell-check-dict] if you want to add some words to the spell check dictionary.

[commit-guidelines]: https://www.conventionalcommits.org/en/v1.0.0/
[spell-check-dict]: https://github.com/tier4/autoware-spell-check-dict#how-to-contribute

### CODE REVIEW

**Comment by TakaHoribe** [comment]
@takayuki5168 Some questions in bold.

- make calculation cost low
    - Fixing out curve results in low calculation cost since solvability gets higher.
        - **want more description**
    - use sparse matrix
        - **Is it to enable the warm start?**
    - use Linf norm for MPT's soft constraints instead of L2 norm
        - **Is there any bad side effects?**
    - make size of obstacle clearance map small
        - **want more description**
    - connect optimized points and remain points with spline interpolation instead of non-robust EB
        - **So, in the previous, EB (for smoothing) -> MPT -> EB (for coupling), but now EB (for smoothing) -> MPT -> spline (for coupling). Am I right?**
- make optimization stable (by dealing with the fixed length well)
    - make fixing distance not so long, and remove them from optimization (previously, fixing points are also optimization's variable)
        -  **how many points are used?**

File: planning/obstacle_avoidance_planner/config/obstacle_avoidance_planner.param.yaml [hunk 1]
```diff
@@ -1,99 +1,164 @@
...
+
+      common:
+        num_curvature_sampling_points: 5 # number of sampling points when calculating curvature
+
+        delta_arc_length_for_mpt_points: 0.5 # delta arc length when generating mpt [m]
+
+      kinematics:
+        max_steer_deg: 40.0 # max steering angle [deg]
+
+        # By commenting this, we can run vehicle-centered optimization for MPT
```
Comment (TakaHoribe) [inline planning/obstacle_avoidance_planner/config/obstacle_avoidance_planner.param.yaml:110]: what do you mean?

**Comment by yukkysaito** [comment]
pls change merge  branch to main

**Comment by takayuki5168** [comment]
@TakaHoribe Give me LGTM~

**Review by TakaHoribe — APPROVED** [review]
L
G
T
M
!!!

**Pull request merged**